### PR TITLE
Add bit injection

### LINKED ISSUE #4590: ABY: Bit injection
## Feature Description
Implement the bit injection algorithm described in [here](https://eprint.iacr.org/2018/403.pdf)
**EDIT**: We can defer this to later since the local party ("me") is the one that orchestrates all the computation and it plays a crypto provider kind of role.

## Is your feature request related to a problem?
The Falcon protocol requires the bit injection algorithm

### PR DESCRIPTION
**Pull request by gmuraru**
## Description
Added the bit injection for changing a share from ring_size 2 (binary) to an arithmetic share.
Followed the explanations from [here](https://eprint.iacr.org/2018/403.pdf).

Currently, we have a higher communication cost because we use another party that orchestrates all the communication between the parties.

Another idea (will be more readable, but in a real-life scenario will require more communication cost), is to make the remote workers reshare their binary shares and do the same computation -- this will also require adding serialization for the RST (a good next step might be this)

Fixes #4590 - this is for the **Malicious Setup**. Will start working on the Semi-Honest one - That one requires an **Oblivious Transfer Component** to it

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        shares_b_1 = dict(zip(players, (b1_share_p1, b1_share_p2, b1_share_p3)))
+        shares_b_2 = dict(zip(players, (b2_share_p1, b2_share_p2, b2_share_p3)))
+        shares_b_3 = dict(zip(players, (b3_share_p1, b3_share_p2, b3_share_p3)))
+
+        rst1 = ReplicatedSharingTensor(shares=shares_b_1)
+        rst2 = ReplicatedSharingTensor(shares=shares_b_2)
+        rst3 = ReplicatedSharingTensor(shares=shares_b_3)
+
+        tmp = rst1 + rst2 - 2 * rst1 + rst2
+        res = tmp + rst3 - 2 * rst1 + rst2
```
Comment (gmuraru) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:63]: BUG HERE

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+        return_wrapper = rst_binary.is_wrapper
+        if return_wrapper:
+            rst_binary = rst_binary.child
+
+        players = rst_binary.players
+        shares_p1, shares_p2, shares_p3 = rst_binary.retrieve_pointers()
+        zeros = torch.zeros(shares_p1[0].shape, dtype=torch.long)
+
+        zeros_p1 = zeros.send(players[0])
```
Comment (gmuraru) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:18]: What else we can do here:
zero_share1 = gen_alpha3of3(players[0])
zero_share2 = gen_alpha3of3(players[1])
zero_share3 = gen_alpha3of3(players[2])
```
Where:
```zero_share1 + zero_share2 + zero_share3 = 0``` (using the corelated randomness)

And then we will have:
```
x1_shares = ((x_b_1, zero_share1), (zero_share2, zero_share3), (zero_share3, x_b_1))
x2_shares = ((zero_share1, x_b_2), (x_b_2, zero_share3), (zero_share3, zero_share1))
x3_shares = ((zero_share1, zero_share2), (zero_share2, x_b_3), (x_b_3, zero_share1))
```

The end result will be computed the same, only that we use the ```correlated randomness``` (also this require sending the zero_share1, zero_share2, zero_share3 to the other parties -- more communication overhead)

Comment (gmuraru) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:18]: What else we can do here.
Using the correlated randomness generate:
(val1, val2) = gen_alpha2of3(player[0])
(val2, val3) = gen_alpha2of3(player[1])
(val3, val1) = gen_alpha2of3(player[2])

zero_share1 = val1 ^ val2
zero_share2 = val2 ^ val3
zero_share3 = val3 ^ val1
```
Where:
```zero_share1 xor zero_share2 xor zero_share3 == 0```

After that we will have:
```
x1_shares = ((x_b_1, zero_share2), (zero_share2, zero_share3), (zero_share3, x_b_1))
x2_shares = ((zero_share1, x_b_2), (x_b_2, zero_share3), (zero_share3, zero_share1))
x3_shares = ((zero_share1, zero_share2), (zero_share2, x_b_3), (x_b_3, zero_share1))
```
And at the end we will have the same result:
(x1, x2, x3 original shares in ring size2, xor can also be seen as a + here since we are in ring size 2)
Rec(x1, x2, x3) = [[x1 xor x2 xor x3]] = d + x3 - 2 * d * x3 where d = x1 + x2 - 2 * x1 * x2 (using the arithmetic shares)

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        players = rst_binary.players
+        shares_p1, shares_p2, shares_p3 = rst_binary.retrieve_pointers()
+
+        """
+            x shared in ring size 2
+            x_b = (x_b_1, x_b_2, x_b_3)
+
+            Share the shares in ring size L
+            Resharig the binary shars in arithmetic ones should have no communication cost
+            but because we use an orchestrator there is added the communicatio overhead
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:23]:
```suggestion
            Resharing the binary shares in arithmetic ones should have no communication cost
            but because we use an orchestrator there is added communication overhead
```

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+            x1_shares = ((x_b_1, 0), (0, 0), (0, x_b_1))
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:25]: Can you start from 0 instead of 1? :D 
it will be simpler for when you combine with ....[0]

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            x2_shares = ((0, x_b_2), (x_b_2, 0), (0, 0))
+            x3_shares = ((0, 0), (0, x_b_3), (x_b_3, 0))
+
+            For emulating a xor on arithmetic shares:
+                * [[v1 xor v2]] = [[v1]] + [[v2]] - 2 * [[v1]] * [[v2]]
+
+            (x1_shares xor x2_shares xor x3_shares) == x_b (ring size L)
+        """
+
+        zero_share_p1 = gen_alpha_3of3(players[0]).wrap()
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:35]: stupid question: but is the size of the output ring (L) fixed or should it be configurable?

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        """
+        Since we shared the seeds at the beginning we know that
+        zero_share_p1 + zero_share_p2 + zero_share_p3 == 0
+        """
+
+        b1_share_p1 = (shares_p1[0] + zero_share_p1, zero_share_p2.copy().move(players[0]))
+        b1_share_p2 = (zero_share_p2, zero_share_p3.copy().move(players[1]))
+        b1_share_p3 = (zero_share_p3, shares_p3[1] + zero_share_p1.copy().move(players[2]))
+
+        b2_share_p1 = (zero_share_p1.copy(), shares_p1[1] + zero_share_p2.copy().move(players[0]))
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:48]: not sure you need the copy() here:
`(zero_share_p1.copy(), sh`  -> `(zero_share_p1, sh`

Comment (LaRiffle) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:48]: I don't think you can reuse randomness for the three shared values: this might be a security issue.

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+        shares_b_1 = dict(zip(players, (b1_share_p1, b1_share_p2, b1_share_p3)))
+        shares_b_2 = dict(zip(players, (b2_share_p1, b2_share_p2, b2_share_p3)))
+        shares_b_3 = dict(zip(players, (b3_share_p1, b3_share_p2, b3_share_p3)))
+
+        rst1 = ReplicatedSharingTensor(shares=shares_b_1)
+        rst2 = ReplicatedSharingTensor(shares=shares_b_2)
+        rst3 = ReplicatedSharingTensor(shares=shares_b_3)
+
+        tmp = rst1 + rst2 - 2 * rst1 * rst2
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:64]: You might have an easy win on syntax simplicity % ABY3 by implementing xor as a method of RST?

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py (hunk 6 above)
Comment (LaRiffle) [reply syft/frameworks/torch/mpc/aby3/aby3_helper.py:48]: ah maybe not... no you're right

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        b2_share_p2 = (shares_p2[0] + zero_share_p2.copy(), zero_share_p3.copy().move(players[1]))
+        b2_share_p3 = (zero_share_p3.copy(), zero_share_p1.copy().move(players[2]))
+
+        b3_share_p1 = (zero_share_p1.copy(), zero_share_p2.copy().move(players[0]))
+        b3_share_p2 = (zero_share_p2.copy(), shares_p2[1] + zero_share_p3.copy().move(players[1]))
+        b3_share_p3 = (shares_p3[0] + zero_share_p3.copy(), zero_share_p1.copy().move(players[2]))
+
+        shares_b_1 = dict(zip(players, (b1_share_p1, b1_share_p2, b1_share_p3)))
+        shares_b_2 = dict(zip(players, (b2_share_p1, b2_share_p2, b2_share_p3)))
+        shares_b_3 = dict(zip(players, (b3_share_p1, b3_share_p2, b3_share_p3)))
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:58]: I wonder if all this work can't be simplified to be slightly more compact? 🤔 [not super important but just wondering]

**Review by LaRiffle — CHANGES_REQUESTED** [review]

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py (hunk 7 above)
Comment (gmuraru) [reply syft/frameworks/torch/mpc/aby3/aby3_helper.py:64]: Yep - currently it is implemented in Falcon - I think it might be better to move it in ABY3, since Falcon builds on top of ABY3.

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py (hunk 8 above)
Comment (gmuraru) [reply syft/frameworks/torch/mpc/aby3/aby3_helper.py:58]: I think in the RST we can have a method refresh (like in AST) where we reshare the values? and we will simply do.
```
arithmetic_share_1 = shares_p1[0].reshare() <--- and here do all the stuff I did in this function?
```

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py (hunk 6 above)
Comment (gmuraru) [reply syft/frameworks/torch/mpc/aby3/aby3_helper.py:48]: Yep - will change. Also, wanted to ask this in today's meeting

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py (hunk 5 above)
Comment (gmuraru) [reply syft/frameworks/torch/mpc/aby3/aby3_helper.py:35]: It should be configurable depending in what ring_size we want to change --- I have a TODO in the Description - I need to add that :D

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py (hunk 4 above)
Comment (gmuraru) [reply syft/frameworks/torch/mpc/aby3/aby3_helper.py:25]: Sure

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py (hunk 7 above)
Comment (LaRiffle) [reply syft/frameworks/torch/mpc/aby3/aby3_helper.py:64]: this tick is vey standard, you could really add a method to RST like
```python
def __xor__(self, other):
    return self + other - 2 * self * other
```

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py (hunk 8 above)
Comment (LaRiffle) [reply syft/frameworks/torch/mpc/aby3/aby3_helper.py:58]: yep probably!

**Comment by aanurraj** [comment]
sorry i accidentally closed :P

**Comment by gmuraru** [comment]
> sorry i accidentally closed :P

No worries :D - I have to add one more thing to this

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py (hunk 7 above)
Comment (gmuraru) [reply syft/frameworks/torch/mpc/aby3/aby3_helper.py]: Yep - we had it in the ReplicatedSharingTensor and did not want to use it since Falcon builds on top of what Aby is doing

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+            Step 3: x3_shares = ((0, 0), (0, x_b_2), (x_b_2, 0))
+
+            For emulating a xor on arithmetic shares:
+                * [[v1 xor v2]] = [[v1]] + [[v2]] - 2 * [[v1]] * [[v2]]
+
+            (x0_shares xor x1_shares xor x2_shares) == x_b (ring size L)
+        """
+
+        # Step 1
+        rst1 = FalconHelper.convert_b2a((shares_p0[0], shares_p2[1]), players)
```
Comment (gmuraru) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:37]: ```convert_b2a``` should stay in something that is more "Framework agnostic" but for the time being, I think it is ok to be in FalconHelper since we use ```ReplicatedSharingTensor```

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+        # Step 2
+        rst2 = FalconHelper.convert_b2a((shares_p1[0], shares_p0[1]), players)
+
+        # Step 3
+        rst3 = FalconHelper.convert_b2a((shares_p2[0], shares_p1[1]), players)
+
+        tmp = FalconHelper.xor(rst1, rst2)
```
Comment (gmuraru) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:45]: Also the ```xor``` should be moved to something more agnostic.
Maybe have a file name ```smpc_utils.py```, but for the moment I think this is fine.

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
+from syft.frameworks.torch.mpc.falcon.falcon_helper import FalconHelper
+
+
+class ABY3Helper:
+    def bit_inject(rst_binary):
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:5]: can you add typing info for rst_binary? (ReplicatedSharingTensor?)

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py (hunk 9 above)
Comment (LaRiffle) [reply syft/frameworks/torch/mpc/aby3/aby3_helper.py:34]: Yeah this looks fine!

File: syft/frameworks/torch/mpc/falcon/falcon_helper.py [hunk 12]
```diff
@@ -23,6 +28,50 @@ def __public_unfold(image, kernel_size, padding):
...
+        arithmetic_shares = dict(zip(players, (share_p0, share_p1, share_p2)))
+        arithmetic_shares[player0][0] += shares[0]
+        arithmetic_shares[player1][1] += shares[1]
+
+        shares = {key: tuple(val) for key, val in arithmetic_shares.items()}
+
+        return ReplicatedSharingTensor(shares=shares)
+
     @staticmethod
     def xor(
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/falcon/falcon_helper.py:76]: I really think the `__xor__` method shoud be added to the ReplicatedSharingTensor and not to the FalconHelper

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 13]
```diff
@@ -87,14 +92,13 @@ def retrieve_shares(self):
...
-            shares.append(pointer.copy().get())
+            shares.append(pointer[0].copy().get())
         return shares
 
     def retrieve_pointers(self):
         shares_map = self.__get_shares_map()
         players = list(shares_map.keys())
-        pointers = list(shares_map[players[0]])
-        pointers.append(shares_map[players[1]][1])
+        pointers = [shares_map[player] for player in players]
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:101]: Maybe you can simply put here: `pointers = shares_map.values()` ?

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 14]
```diff
@@ -352,6 +358,8 @@ def simplify(worker: AbstractWorker, tensor: "ReplicatedSharingTensor") -> tuple
 
         tensor.set_garbage_collect_data(False)
 
+        tensor.set_garbage_collect_data(False)
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:361]: Duplicate line?

File: syft/frameworks/torch/mpc/aby3/aby3_helper.py (hunk 11 above)
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/aby3/aby3_helper.py:5]: Can you add to the signature an argument `ring_size` defaulting to 2 ** 32 that you propagate to `convert_b2a` ?

**Review by LaRiffle — CHANGES_REQUESTED** [review]
Some small details but I guess we're almost there! :)

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 13 above)
Comment (gmuraru) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:101]: In python3.8 (I am not sure if this is also in python3.7):
```
>>> a = {'a': 3, 'b': 4}
>>> a
{'a': 3, 'b': 4}
>>> a.values()
dict_values([3, 4]) <--- those are dict values
```

Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:101]: Ok, maybe `list(shares_map.values())`  to be sure?
